homeassistant/components/vicare/fan.py

```diff
@@ -0,0 +1,192 @@
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ServiceValidationError
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util.percentage import (
+    ordered_list_item_to_percentage,
+    percentage_to_ordered_list_item,
+)
+
+from .const import DEVICE_LIST, DOMAIN
+from .entity import ViCareEntity
+from .types import VentilationMode, VentilationProgram, ViCareDevice
+
+_LOGGER = logging.getLogger(__name__)
+
+ORDERED_NAMED_FAN_SPEEDS = [
+    VentilationProgram.LEVEL_ONE,
+    VentilationProgram.LEVEL_TWO,
+    VentilationProgram.LEVEL_THREE,
+    VentilationProgram.LEVEL_FOUR,
+]
+
+PRESET_MODES = [
+    VentilationMode.PERMANENT,
+    VentilationMode.VENTILATION,
+    VentilationMode.SENSOR_DRIVEN,
+    VentilationMode.SENSOR_OVERRIDE,
+]
+
+
+def _build_entities(
+    device_list: list[ViCareDevice],
+) -> list[ViCareFan]:
+    """Create ViCare fan entities for a device."""
+
+    return [
+        ViCareFan(device.config, device.api, "ventilation")
+        for device in device_list
+        if isinstance(device.api, PyViCareVentilationDevice)
+    ]
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the ViCare fan platform."""
+
+    device_list = hass.data[DOMAIN][config_entry.entry_id][DEVICE_LIST]
+
+    async_add_entities(
+        await hass.async_add_executor_job(
+            _build_entities,
+            device_list,
+        )
+    )
+
+
+class ViCareFan(ViCareEntity, FanEntity):
+    """Representation of the ViCare ventilation device."""
+
+    _attr_speed_count = len(ORDERED_NAMED_FAN_SPEEDS)
+    _attr_supported_features = FanEntityFeature.SET_SPEED | FanEntityFeature.PRESET_MODE
+    _attributes: dict[str, Any] = {
+        "active_vicare_mode": None,
+        "active_vicare_program": None,
+        "vicare_modes": None,
+        "vicare_programs": None,
+    }
+
+    def __init__(
+        self,
+        device_config: PyViCareDeviceConfig,
+        device: PyViCareDevice,
+        translation_key: str,
+    ) -> None:
+        """Initialize the fan entity."""
+        super().__init__(device_config, device, translation_key)
+        self._attr_translation_key = translation_key
+
+        with suppress(PyViCareNotSupportedFeatureError):
+            self._attributes["vicare_modes"] = self._api.getAvailableModes()
+        self._attributes["vicare_programs"] = self._api.getAvailablePrograms()
+
+    def update(self) -> None:
+        """Update state of fan."""
+        try:
+            with suppress(PyViCareNotSupportedFeatureError):
+                self._attributes["active_vicare_mode"] = self._api.getActiveMode()
+            with suppress(PyViCareNotSupportedFeatureError):
+                self._attributes["active_vicare_program"] = self._api.getActiveProgram()
```

Comment:
During my test I noticed that the API sometimes needs several minutes to deliver the updated value after changing the fan speed. Is this due to the limitation of Viessmann or the implementation of the Vicare library?

Reply:
The API response cache is invalidated after a *set* command. This should pull the newest data from the API. But it's known on other occasions that the API is not updating that fast:

https://github.com/oischinger/ha_vicare/blob/8efbd8ec8b20e8cf816be46df7ca843de02fd316/custom_components/vicare/switch.py#L134-L136

Reply:
Probably we would need to have some kind of blackout period where the entity reads the values not from the api but *remembers* that it was set to a certain value.